src/vmm/benches/main.rs

```diff
@@ -0,0 +1,188 @@
+    mut snapshot_reader: &[u8],
+    snapshot_len: usize,
+    vm: VersionMap,
+    crc: bool,
+) {
+    if crc {
+        Snapshot::load::<&[u8], MicrovmState>(&mut snapshot_reader, snapshot_len, vm).unwrap();
+    } else {
+        Snapshot::unchecked_load::<&[u8], MicrovmState>(&mut snapshot_reader, vm).unwrap();
+    }
+}
+
+#[inline]
+pub fn bench_create_snapshot<W: std::io::Write>(
+    mut snapshot_writer: &mut W,
+    vm: VersionMap,
+    crc: bool,
+    state: &mut MicrovmState,
+) {
+    let mut snapshot = Snapshot::new(vm.clone(), vm.latest_version());
+
+    if crc {
+        snapshot.save(&mut snapshot_writer, state).unwrap();
+    } else {
+        snapshot
+            .save_without_crc(&mut snapshot_writer, state)
+            .unwrap();
+    }
+}
+
+fn create_microvm_state(is_diff: bool) -> MicrovmState {
+    let snapshot_file = TempFile::new().unwrap();
+    let memory_file = TempFile::new().unwrap();
+
+    let pid = unsafe { libc::fork() };
+    match pid {
+        0 => {
+            set_panic_hook();
+
+            let (vmm, _) = create_vmm(Some(NOISY_KERNEL_IMAGE), is_diff);
+
+            // Be sure that the microVM is running.
+            thread::sleep(Duration::from_millis(200));
+
+            // Pause microVM.
+            vmm.lock().unwrap().pause_vm().unwrap();
+
+            // Create snapshot.
+            let snapshot_type = match is_diff {
+                true => SnapshotType::Diff,
+                false => SnapshotType::Full,
+            };
+            let snapshot_params = CreateSnapshotParams {
+                snapshot_type,
+                snapshot_path: snapshot_file.as_path().to_path_buf(),
+                mem_file_path: memory_file.as_path().to_path_buf(),
+                version: Some(String::from("0.24.0")),
+            };
+
+            {
+                let mut locked_vmm = vmm.lock().unwrap();
+                persist::create_snapshot(&mut locked_vmm, &snapshot_params, VERSION_MAP.clone())
+                    .unwrap();
+            }
+
+            vmm.lock().unwrap().stop(0);
+        }
+        vmm_pid => {
+            // Parent process: wait for the vmm to exit.
+            wait_vmm_child_process(vmm_pid);
+        }
+    }
+
+    // Deserialize the microVM state from `snapshot_file`.
+    let snapshot_path = snapshot_file.as_path().to_path_buf();
+    let snapshot_file_metadata = std::fs::metadata(snapshot_path).unwrap();
+    let snapshot_len = snapshot_file_metadata.len() as usize;
+    let microvm_state: MicrovmState = Snapshot::load(
+        &mut snapshot_file.as_file(),
+        snapshot_len,
+        VERSION_MAP.clone(),
+    )
+    .unwrap();
+
+    microvm_state
+}
+
+pub fn criterion_benchmark(c: &mut Criterion) {
+    let version_map = VERSION_MAP.clone();
+
+    // Create the microvm state
+    let mut state = create_microvm_state(false);
+
+    // Setup benchmarking with CRC
+    let mut snapshot_state_with_crc = vec![0u8; 1024 * 1024 * 128];
```

Comment:
If we get the snapshot file we do not need to do this.

Reply:
You mean to use the snapshot file that can be obtained above directly for deserialization?

Reply:
Yes, deserialization only.

Reply:
Understood.

Reply:
Discussed with @sandreim and this will remain as it. The reason is that we want to avoid IO operations when doing performance tests and byte buffers are a better option here.